### PR TITLE
ORC-1008: Fix overflow detection code for C++ int64_t / java long

### PR DESCRIPTION
**Pull request by guiyanakuang**
### What changes were proposed in this pull request?

https://github.com/apache/orc/blob/6da96bb8ceb64528d082974efed411c4c29f3408/c%2B%2B/src/Statistics.cc#L180-L190
A counter-example can easily be given
Assume `sum=1`, `update(std::numeric_limits<int64_t>::max(), 3);`
`value * repetitions + _stats.getSum()` is overflowed, but is still a positive number : 9223372036854775806

This pr aims to fix overflow detection code for C++ int64_t / java long.
> ORC-338 Workaround C++ compiler bug in xcode 9.3 by removing an inline function.

As I fixed the implementation. The current update function can be inline.

### Why are the changes needed?

Fix bug.

### How was this patch tested?

Pass the CIs.

### CODE REVIEW

**Comment by guiyanakuang** [comment]
```cmd
CUSTOMBUILD : error : downloading 'ftp://cygwin.osuosl.org/pub/cygwin/noarch/release/tzdata/tzdata-2020e-1.tar.xz' failed [C:\projects\orc\build\tzdata_ep.vcxproj]
           status_code: 78
           status_string: "Remote file not found"
           log:
           --- LOG BEGIN ---
           timeout on name lookup is not supported
```
https://cygwin.osuosl.org/noarch/release/tzdata/
Hi @dongjoon-hyun. It looks like the archive tzdata-2020e-1.tar.xz has been deleted and tzdata-2021b-1-src.tar.xz was created two days ago, do we need to update this third party toolchain.
image

**Comment by pgaref** [comment]
Thanks for the PR @guiyanakuang  -- can we please add a test demonstrating the issue?

**Comment by guiyanakuang** [comment]
> Thanks for the PR @guiyanakuang -- can we please add a test demonstrating the issue?

You're right, I'll add test cases later. : )

File: c++/src/Statistics.hh [hunk 1]
```diff
@@ -964,7 +964,23 @@ namespace orc {
       _stats.setSum(sum);
     }
 
-    void update(int64_t value, int repetitions);
+    void update(int64_t value, int repetitions) {
+      _stats.updateMinMax(value);
+
+      if (_stats.hasSum()) {
+        if (repetitions > 1) {
+          _stats.setHasSum(__builtin_mul_overflow(value, repetitions, &value) == 0);
```
[the remark this answers is not in the document]
Comment (guiyanakuang) [reply c++/src/Statistics.hh:972]: @wgtmac Sorry, I'm still a C++ novice. Let me see how compatible

File: java/core/src/test/org/apache/orc/TestColumnStatistics.java [hunk 2]
```diff
@@ -50,6 +51,16 @@
  */
 public class TestColumnStatistics {
 
+  @Test
+  public void testLongSumOverflow() {
+    TypeDescription schema = TypeDescription.createInt();
+    ColumnStatisticsImpl stats = ColumnStatisticsImpl.create(schema);
+    stats.updateInteger(1, 1);
+    assertTrue(((IntegerColumnStatistics) stats).isSumDefined());
+    stats.updateInteger(Long.MAX_VALUE, 3);
```
Comment (dongjoon-hyun) [inline java/core/src/test/org/apache/orc/TestColumnStatistics.java:60]: Thank you for adding this.

File: c++/src/Statistics.hh (hunk 1 above)
[the remark this answers is not in the document]
Comment (guiyanakuang) [reply c++/src/Statistics.hh:972]: https://docs.microsoft.com/en-us/windows/win32/api/intsafe/nf-intsafe-longadd
https://docs.microsoft.com/en-us/windows/win32/api/intsafe/nf-intsafe-longmult
```c++
#define __builtin_mul_overflow LongMult
#define __builtin_add_overflow LongAdd
```
@wgtmac Can you give me some advice, I have very little practice in C++. I don't have a Win machine to hand, so I can't verify if this is possible at the moment.

[the remark this answers is not in the document]
Comment (guiyanakuang) [reply c++/src/Statistics.hh]: @wgtmac I made some attempts, compatibility with various compilers seemed difficult and I switched to rewriting to C++ following the JDK's algorithm.

File: c++/include/orc/orc-config.hh.in [hunk 3]
```diff
@@ -75,4 +75,38 @@
...
+     * but this method makes the assumption that repetitions > 1
+     */
+    static bool multiplyExact(int64_t value, int64_t repetitions, int64_t* result) {
+      int64_t r = value * repetitions;
+      if (((value < 0 ? -value : value) | repetitions) >> 31 != 0 && r / repetitions != value) {
+        return false;
+      }
+      *result = r;
+      return true;
+    }
```
Comment (guiyanakuang) [inline c++/include/orc/orc-config.hh.in:93]: Hi @dongjoon-hyun @wgtmac. Finally passed the test. But it needs a strict review.
The window environment overflow check I've mimicked the JDK implementation. If there is a better way, please let me know.
In fact it doesn't work well on clang (the compiler has some unknown behaviour that I don't know about)

**Review by dongjoon-hyun — COMMENTED** [review]
From my side, it looks okay. We need @wgtmac 's sign-off.

File: c++/include/orc/orc-config.hh.in (hunk 3 above)
Comment (wgtmac) [reply c++/include/orc/orc-config.hh.in:93]: `#ifdef _MSC_VER` is enough to assert that it is on Windows. You may also leverage cmake_modules/CheckSourceCompiles.cmake to figure out if the gcc builtin functions are available or not. BTW, orc-config.hh.in is to address capability of older c++ standards. If would be better to handle these kinds of functions in Adaptor.hh.in.

**Review by wgtmac — COMMENTED** [review]
Sorry for my late reply. Thanks for the investigation @guiyanakuang and @dongjoon-hyun for the review. I am fine with the impl from jdk. I have left one comment. Any thoughts?

**Comment by guiyanakuang** [comment]
> Sorry for my late reply. Thanks for the investigation @guiyanakuang and @dongjoon-hyun for the review. I am fine with the impl from jdk. I have left one comment. Any thoughts?

@wgtmac. Good advice, I've pushed for a new commit. How do you feel about it?

**Review by wgtmac — APPROVED** [review]
LGTM +1. Thanks @guiyanakuang !

**Comment by guiyanakuang** [comment]
Many thanks to @wgtmac , @dongjoon-hyun  and @pgaref  for reviews and suggestions.  I also learned a lot about C++ in the process of implementing this pr, and it was great to contribute to open source and improve myself. 😄

**Comment by dongjoon-hyun** [comment]
Thank you so much, @guiyanakuang and @wgtmac !

**Comment by guiyanakuang** [comment]
@dongjoon-hyun @wgtmac A small query, do you think this fix needs to be merged into 1.6 and 1.7?

**Comment by dongjoon-hyun** [comment]
At least 1.7?

**Comment by guiyanakuang** [comment]
Yes, because the sum in the statistics may be wrong and the user may take it as the correct conclusion. Do you think this is Major or Minor bug?

**Comment by dongjoon-hyun** [comment]
I believe we need backports.

**Comment by guiyanakuang** [comment]
I think so too

**Comment by xndai** [comment]
We may just need to do this -

*result = x * y;
return (x == 0 || y == 0 || *result / x == y);

**Comment by guiyanakuang** [comment]
> We may just need to do this -
> 
> *result = x * y; return (x == 0 || y == 0 || *result / x == y);

`if (((value < 0 ? -value : value) | repetitions) >> 31 != 0 && r / repetitions != value)`
The current is a copy of the JDK implementation, which intends that values in the range [-2^31, 2^31] can be short-circuited to determine that no overflow will occur. It is also assumed that this range is used more frequently.

Thanks to @xndai for the advice, so I've done some simple benchmarking in https://quick-bench.com/.

They perform about the same under higher versions of the compiler, and it seems that the current implementation is better under lower versions. 

BM_A:  current impl
BM_B:  advice impl
BM_C:  __builtin_mul_overflow

image
image
image



```c++
static bool multiplyExactA(int64_t value, int64_t repetitions, int64_t* result) {
  int64_t r = value * repetitions;
  if (((value < 0 ? -value : value) | repetitions) >> 31 != 0 && r / repetitions != value) {
    return false;
  }
  *result = r;
  return true;
}

static bool multiplyExactB(int64_t x, int64_t y, int64_t* result) {
  int64_t r =  x * y;
  if (x == 0 || y == 0 || r / y == x) {
    *result = r;
    return true;
  }
  return false;
}

static void BM_A(benchmark::State& state) {
  auto x = state.range(0);
  int64_t value;
  for (auto _ : state) {
    for (int64_t i = std::numeric_limits<int64_t>::min(); i < std::numeric_limits<int64_t>::max(); i += 1) {
      for (int64_t j = 2; j <= x; j += 1) {
        multiplyExactA(i, j, &value);
      }
    }
  }
}

static void BM_B(benchmark::State& state) {
  auto x = state.range(0);
  int64_t value;
  for (auto _ : state) {
    for (int64_t i = std::numeric_limits<int64_t>::min(); i < std::numeric_limits<int64_t>::max(); i += 1) {
      for (int64_t j = 2; j <= x; j += 1) {
        multiplyExactB(i, j, &value);
      }
    }
  }
}

static void BM_C(benchmark::State& state) {
  auto x = state.range(0);
  int64_t value;
  for (auto _ : state) {
    for (int64_t i = std::numeric_limits<int64_t>::min(); i < std::numeric_limits<int64_t>::max(); i += 1) {
      for (int64_t j = 2; j <= x; j += 1) {
        __builtin_mul_overflow(i, j, &value);
      }
    }
  }
}

BENCHMARK(BM_A)->Arg(512);

BENCHMARK(BM_B)->Arg(512);

BENCHMARK(BM_C)->Arg(512);
```

**Pull request merged**